src/bin/pg_autoctl/keeper.c

```diff
@@ -1431,6 +1432,208 @@ keeper_register_and_init(Keeper *keeper, NodeState initialState)
 }
 
 
+/*
+ * keeper_register_again registers the given node again to a given monitor URI,
+ * possibly new. This function has been designed to be used from the "enable
+ * monitor" command, in such a scenario:
+ *
+ *   $ pg_autoctl disable monitor --force
+ *   $ pg_autoctl enable monitor --monitor postgresql://...
+ *
+ * The idea is that we have lost the monitor, and we want to re-register nodes
+ * to the new empty monitor, without having to stop pg_autoctl nor Postgres.
+ */
+bool
+keeper_register_again(Keeper *keeper)
+{
+	Monitor *monitor = &(keeper->monitor);
+	KeeperConfig *config = &(keeper->config);
+	PostgresSetup *pgSetup = &(config->pgSetup);
+
+	MonitorAssignedState assignedState = { 0 };
+	ConnectionRetryPolicy retryPolicy = { 0 };
+
+	bool registered = false;
+
+	(void) pgsql_set_monitor_interactive_retry_policy(&retryPolicy);
+
+	/* fetch local metadata for the registration (system_identifier) */
+	if (!pgsql_get_postgres_metadata(&(keeper->postgres.sqlClient),
+									 &pgSetup->is_in_recovery,
+									 keeper->postgres.pgsrSyncState,
+									 keeper->postgres.currentLSN,
+									 &(pgSetup->control)))
+	{
+		log_error("Failed to get the local Postgres metadata");
+		return false;
+	}
+
+	NodeState initialState =
+		pgSetup->is_in_recovery ? WAIT_STANDBY_STATE : SINGLE_STATE;
```

Comment:
There might be a primary and a demoted primary when the monitor is enabled. How does the initialState affect things?

Reply:
You are pointing exactly to why we don't have a fully automated way to register nodes again to a new monitor. To solve the problem, we ask the user to be careful about the node registration ordering. Once a primary has been registered, the other primary might be more interesting to support.

I suppose on this case we are back to the previous situation where the current state needs to be dropped and a new node created from scratch again:

```
$ pg_autoctl stop
$ pg_autoctl drop node --pgdata ...
$ pg_autoctl create postgres --pgdata ... ...
```

But that's only when you're in the unfortunate position of losing the monitor node just after having lost a primary node and then you now have all the nodes back online and you're dealing with all the pieces.